nalgebra-sparse/src/csr.rs

```diff
@@ -184,62 +185,33 @@ impl<T> CsrMatrix<T> {
         num_rows: usize,
         num_cols: usize,
         row_offsets: Vec<usize>,
-        col_indices: Vec<usize>,
-        values: Vec<T>,
+        mut col_indices: Vec<usize>,
+        mut values: Vec<T>,
     ) -> Result<Self, SparseFormatError>
     where
-        T: Scalar,
+        T: Scalar + Zero,
     {
-        use SparsityPatternFormatError::*;
-        let count = col_indices.len();
-        let mut p: Vec<usize> = (0..count).collect();
-
-        if col_indices.len() != values.len() {
-            return Err(SparseFormatError::from_kind_and_msg(
-                SparseFormatErrorKind::InvalidStructure,
-                "Number of values and column indices must be the same",
-            ));
-        }
-
-        if row_offsets.len() == 0 {
-            return Err(SparseFormatError::from_kind_and_msg(
-                SparseFormatErrorKind::InvalidStructure,
-                "Number of offsets should be greater than 0",
-            ));
-        }
-
-        for (index, &offset) in row_offsets[0..row_offsets.len() - 1].iter().enumerate() {
-            let next_offset = row_offsets[index + 1];
-            if next_offset > count {
-                return Err(SparseFormatError::from_kind_and_msg(
-                    SparseFormatErrorKind::InvalidStructure,
-                    "No row offset should be greater than the number of column indices",
-                ));
-            }
-            if offset > next_offset {
-                return Err(NonmonotonicOffsets).map_err(pattern_format_error_to_csr_error);
-            }
-            p[offset..next_offset].sort_by(|a, b| {
-                let x = &col_indices[*a];
-                let y = &col_indices[*b];
-                x.partial_cmp(y).unwrap()
-            });
-        }
-
-        // permute indices
-        let sorted_col_indices: Vec<usize> =
-            Vec::from_iter((p.iter().map(|i| &col_indices[*i])).cloned());
-
-        // permute values
-        let sorted_values: Vec<T> = Vec::from_iter((p.iter().map(|i| &values[*i])).cloned());
-
-        return Self::try_from_csr_data(
+        let result = cs::validate_and_optionally_sort_cs_data(
             num_rows,
             num_cols,
-            row_offsets,
-            sorted_col_indices,
-            sorted_values,
+            &row_offsets,
+            &mut col_indices,
+            Some(&mut values),
+            true,
         );
+
+        match result {
+            Ok(()) => unsafe {
+                let pattern = SparsityPattern::from_offset_and_indices_unchecked(
```

Comment:
same as up above, cna probably move the unsafe to only encapsulate the unchecked call.